piccolo_api/mfa/authenticator/tables.py

```diff
@@ -0,0 +1,270 @@
+
+    @classmethod
+    async def create_new(
+        cls,
+        user_id: int,
+        db_encryption_key: str,
+        recovery_code_count: int = 8,
+    ) -> t.Tuple[AuthenticatorSecret, t.List[str]]:
+        """
+        Returns the new ``AuthenticatorSecret`` and the unhashed recovery
+        codes. This is the only time the unhashed recovery codes will be
+        accessible.
+
+        :param user_id:
+            The user to create the secret for.
+        :param db_encryption_key:
+            The secret is encrypted in the database - ``db_encryption_key``
+            can be any reasonably random string. It provides a little extra
+            protection vs storing the secret in plain text.
+        :param recovery_code_count:
+            How many recovery codes to generate for the user - this allows
+            them to still gain access if their phone is lost.
+
+        """
+        # Generate recovery codes
+
+        recovery_codes = [
+            generate_recovery_code() for _ in range(recovery_code_count)
+        ]
+
+        #######################################################################
+        # Hash the recovery codes
+
+        # Use the hashing logic from BaseUser.
+        # We want to use the same salt for all of the user's recovery codes,
+        # otherwise logging in using a recovery code will take a long time.
+        salt = BaseUser.get_salt()
+
+        hashed_recovery_codes = [
+            BaseUser.hash_password(password=recovery_code, salt=salt)
+            for recovery_code in recovery_codes
+        ]
+
+        #######################################################################
+        # Generate a shared secret
+
+        secret = cls.generate_secret()
+
+        # We'll encrypt the secret for storing in the database.
+        encrypted_secret = _encrypt(
+            value=secret, db_encryption_key=db_encryption_key
+        )
+
+        #######################################################################
+
+        instance = cls(
+            {
+                cls.user_id: user_id,
+                cls.secret: encrypted_secret,
+                cls.recovery_codes: hashed_recovery_codes,
+            }
+        )
+        await instance.save()
+
+        return (instance, recovery_codes)
+
+    @classmethod
+    async def revoke_all(cls, user_id: int):
+        now = datetime.datetime.now(tz=datetime.timezone.utc)
+        await cls.update({cls.revoked_at: now}).where(
+            cls.user_id == user_id,
+            cls.revoked_at.is_null(),
+        )
+
+    @classmethod
+    async def authenticate(
+        cls, user_id: int, code: str, db_encryption_key: str
+    ) -> bool:
+        secret = (
+            await cls.objects()
+            .where(
+                cls.user_id == user_id,
+                cls.revoked_at.is_null(),
+            )
+            .order_by(cls.created_at, ascending=False)
+            .first()
+        )
+
```

Comment:
My original intent was to allow the user to register multiple devices, but then it adds complexity, because the user needs a way of identifying the device, if they want to disable it.